docs/design/app/states.md

```diff
@@ -0,0 +1,204 @@
+is still S0. If it has become S1, then we need to throw away whatever we computed in pre-handle, and recompute it
+again, this time with the **current working state**.
+
+Therefore, when the [Pre-handle Workflow](workflows.md#pre-handle-workflow) processes a transaction, it **must** use a
+`ReadableState` that tracks all state that was accessed, and remember the set of state that was accessed.When the
+transaction is handled on the _handle transaction_ thread, we will check whether any state accessed during pre-handle
+has been modified since it was accessed. If so, we throw away any result from pre-handle and run again, this time
+serially on the _handle transaction_ thread on with the **current working state**. If no state used during pre-handle
+has been modified, then we can trust the pre-work we did and complete the transaction.
+
+### States used during Transaction Handling
+
+When a state is used during the [Transaction Handling Workflow](workflows.md#handling-transactions), it is always
+a `WritableState` based on the **current working state**. The only workflow that supports writable state is this one.
+At the end of the `Round`, after **all** transactions in the round have been handled, the state is committed to the
+merkle tree.
+
+## State Registry
+
+When a Service instance is created, it is given a `StateRegistry`. Through this interface, the service is able to manage
+its `State`s. It can create and register new `State`s, and decide whether that state should reside in memory, or on
+disk. It can also migrate state, or even delete state. The `StateRegistry` given to the service instances is scoped to
+that service. It is not possible for one service to look up `State` for another service through the `StateRegistry`.
+Each `State` is referenced by a "state key", which is likewise scoped to the `StateRegistry`. Two different service
+instances could use the same "state key" without naming collisions. Each `StateRegistry` is thus a unique namespace.
+
+The following code is an example of a service which makes use of the `StateRegistry` to create a new `WritableState`. It
+makes use of the `StateDefinition` to do so.
+
+```java
+public class MyService implements Service {
+    public MyService(final StateRegistry registry) {
+        // This call will either _get_ the state associated with MY_STATE, or it will _register_ the new state
+        final var myState = registry.registerOrMigrate(
+                "MY_STATE",
+                (builder, existingStateOpt) -> {
+                    if (existingStateOpt.isEmpty()) {
+                        // This is a genesis condition -- the state does not already exist
+                        return builder.inMemory("MY_STATE")
+                                .define();
+                    } else {
+                        // Not genesis. If I needed to do a migration, I could do it here. For example, I could
```

Comment:
It's not immediately obvious how the `Service` will know if it _needs_ to do a migration, since that will typically depend on the deserialized [`SoftwareVersion`](https://github.com/swirlds/swirlds-platform/blob/develop/swirlds-common/src/main/java/com/swirlds/common/system/SwirldState.java#L69).

Reply:
Good one! We could add the `SoftwareVersion` argument to the callback, and that would tell the service what the current software version is, but how would it know the software version of the existing data (i.e. how does it determine the data it is looking at in the State is v3 while the version of the app is v4 and therefore it needs to upgrade)? What would you suggest?

Reply:
Maybe store the version as part of the data.

Reply:
I have a different proposal we can try out for `StateRegistry`:

```
/**
 * Defines a registry of states for services.
 */
public interface StateRegistry {

    /**
     * Gets the current {@link SoftwareVersion} of this application at the time of startup. This may be different from
     * the {@link #getExistingVersion()} if the system is starting with an older, existing body of state from an
     * older version.
     *
     * @return The version of the current system.
     */
    @NonNull SoftwareVersion getCurrentVersion();

    /**
     * Gets the {@link SoftwareVersion} of the state at the time of startup. This may be different from
     * {@link #getCurrentVersion()} if the system is starting with an older, existing body of state from an older
     * version.
     *
     * @return The version of the system's state that was loaded, or {@link SoftwareVersion#NO_VERSION}.
     */
    @NonNull SoftwareVersion getExistingVersion();

    /**
     * Gets an existing state from the {@link StateRegistry}, identified by {@code stateKey}.
     *
     * @param stateKey The key of the state to get.
     * @return the {@link WritableState} associated with the key, if there is one. This state should <b>NOT</b> be held
     *      but should only be used during construction of the service. It refers to a mutable state, which will not
     *      be mutable once the service is up and running! The returned value may be null if there is no such state.
     * @param <K> The key for the state.
     * @param <V> The value for the state.
     */
    @Nullable
    <K, V> WritableState<K, V> getState(@NonNull String stateKey);

    /**
     * Replaces any existing state associated with {@code stateKey} with a new, empty state as defined by
     * {@link StateDefinition}. Or, create a new state that didn't exist before.
     *
     * <p>If you need to migrate data from an old state, please use {@link #getState(String)}
     * first to get the old state, then replace the state with this method, and then call {@link #getState(String)}
     * again to get the new state. Then migrate data from old to new.
     *
     * @param stateKey The state key. Cannot be null and must be a valid state key.
     * @return A {@link StateDefinition} to be used to define the state.
     */
    StateDefinition defineNewState(@NonNull String stateKey);

    /**
     * Removes the specified state from the registry.
     *
     * @param stateKey The key of the state to remove
     */
    void removeState(@NonNull String stateKey);
}
```